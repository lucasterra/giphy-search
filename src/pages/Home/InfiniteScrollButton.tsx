import React, { useRef, useCallback } from 'react';
import styled from '@emotion/styled';
import { useIntersectionObserver } from '../../hooks/useIntersectionObserver';

function useInfiniteScroll(loadMore: () => void) {
  const ref = useRef<HTMLButtonElement | null>(null);
  const onIntersectionChanged = useCallback(
    (intersecting: boolean) => {
      if (intersecting) {
        loadMore();
      }
    },
    [loadMore]
  );
  useIntersectionObserver(ref, null, '50px', 0.1, onIntersectionChanged);

  return ref;
}

const Button = styled.button({});

export interface InfiniteScrollButtonProps {
  loadMore: () => void;
  hidden: boolean;
  isLoading: boolean;
}

export const InfiniteScrollButton: React.FC<InfiniteScrollButtonProps> = ({
  loadMore,
  hidden,
  isLoading,
  ...props
}) => {
  const ref = useInfiniteScroll(loadMore);

  return (
    <Button
      ref={ref}
      style={hidden ? { display: 'none' } : undefined}
      onClick={loadMore}
      {...props}
    >
      {isLoading ? 'Loading' : 'Load more'}
    </Button>
  );
};
