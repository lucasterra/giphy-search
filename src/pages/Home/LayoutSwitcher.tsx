import React from 'react';
import styled from '@emotion/styled';

export type LayoutMode = 'single_column' | 'three_column';

interface LayoutSwitcherProps {
  setLayoutMode: (layoutMode: LayoutMode) => void;
  layoutMode: LayoutMode;
}

const Wrapper = styled.div({
  display: 'flex',
});

const Button = styled.button({
  flexGrow: 1,
  color: 'white',
});

export const LayoutSwitcher: React.FC<LayoutSwitcherProps> = ({
  setLayoutMode,
  layoutMode,
}) => (
  <Wrapper>
    <Button onClick={() => setLayoutMode('single_column')}>1 column</Button>
    <Button onClick={() => setLayoutMode('three_column')}>3 column</Button>
  </Wrapper>
);
