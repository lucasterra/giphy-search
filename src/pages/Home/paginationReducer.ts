import { Root } from '../../api/GIPHY';

// Update the search input
export const SET_TEXT = 'SET_TEXT';

// Fetch more images with infinite scrolling
export const NEXT_PAGE = 'NEXT_PAGE';

// Set image data, the data returned from GIPHY
export const SET_IMAGE_DATA = 'SET_IMAGE_DATA';

// Actions
interface SetTextAction {
  type: typeof SET_TEXT;
  payload: string;
}

interface NextPageAction {
  type: typeof NEXT_PAGE;
}

interface SetImageDataAction {
  type: typeof SET_IMAGE_DATA;
  payload: Root | null;
}

export type Action = SetTextAction | NextPageAction | SetImageDataAction;

// State
export interface ReducerState {
  text: string;
  page: number;
  isLoading: boolean;
  imageData: Root | null;
}

export const reducer = (state: ReducerState, action: Action) => {
  switch (action.type) {
    case NEXT_PAGE:
      return { ...state, page: state.page + 1, isLoading: true };
    case SET_TEXT:
      return { ...state, page: 0, text: action.payload, isLoading: true };
    case SET_IMAGE_DATA:
      return { ...state, imageData: action.payload, isLoading: false };
    default:
      return state;
  }
};
