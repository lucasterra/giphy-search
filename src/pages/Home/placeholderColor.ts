// Colors used for placeholders while image is being loaded.
// Each image index will use a different color
export const placeholderColors = [
  '#c4b7d2',
  '#9f5a57',
  '#4329ad',
  '#ae2955',
  '#335dad',
  '#60a372',
  '#ad7905',
  '#87cabf',
  '#f2a970',
];

export function getPlaceholderColor(index: number) {
  return placeholderColors[index % placeholderColors.length];
}
