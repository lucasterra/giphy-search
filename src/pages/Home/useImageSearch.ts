import { useEffect, useCallback, useReducer } from 'react';
import { useDebounce } from 'use-debounce';
import { searchImages, Root } from '../../api/GIPHY';
import {
  reducer,
  SET_TEXT,
  NEXT_PAGE,
  SET_IMAGE_DATA,
} from './paginationReducer';

const cache = {};
const useImageFetcher = (data: {
  searchTerm: string;
  count: number;
  offset: number;
  imageData: Root | null;
  setImageData: (data: Root | null, mergeData?: boolean) => void;
}) => {
  const { searchTerm, count, offset, imageData, setImageData } = data;

  useEffect(() => {
    let isFresh = true;
    searchImages(searchTerm, count, offset, cache).then((result) => {
      if (isFresh) {
        if (offset === 0) {
          // first page, just override all the results
          setImageData(result);
        } else {
          // second or more page, merge image data
          setImageData(result, true);
        }
      }
    });

    return () => {
      isFresh = false;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchTerm, count, offset]);

  return imageData;
};

export const useImageSearch = (imagesPerPage: number) => {
  const [reducerData, dispatch] = useReducer(reducer, {
    text: '',
    page: 0,
    isLoading: false,
    imageData: null,
  });
  const { text, isLoading, page, imageData } = reducerData;

  // debounce text changes, to avoid calling GIPHY API on every keystroke
  const [debouncedText] = useDebounce(text, 250);

  // functions passed on hook return
  const setSearchTerm = useCallback((text: string) => {
    dispatch({ type: SET_TEXT, payload: text });
  }, []);

  // called to load more images
  const loadMore = useCallback(() => {
    if (isLoading) {
      return;
    }

    dispatch({ type: NEXT_PAGE });
  }, [isLoading]);

  // update image data object
  const setImageData = useCallback(
    (data: Root | null, mergeData?: boolean) => {
      if (mergeData && imageData !== null && data !== null) {
        dispatch({
          type: SET_IMAGE_DATA,
          payload: { ...data, data: [...imageData.data, ...data.data] },
        });
      } else {
        dispatch({ type: SET_IMAGE_DATA, payload: data });
      }
    },
    [imageData]
  );

  // fetch data from the API
  useImageFetcher({
    searchTerm: debouncedText,
    count: imagesPerPage,
    offset: page * imagesPerPage,
    imageData,
    setImageData,
  });

  // whether we have more images to load or not
  const hasMoreImages = Boolean(
    imageData && imageData.data.length < imageData.pagination.total_count
  );

  return { ...reducerData, setSearchTerm, loadMore, hasMoreImages };
};
