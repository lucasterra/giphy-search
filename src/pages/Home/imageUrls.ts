import { GIFObject } from '../../api/GIPHY/schema';
import { LayoutMode } from './LayoutSwitcher';

export function getMainUrl(img: GIFObject, layoutMode: LayoutMode) {
  return layoutMode === 'single_column'
    ? [
        img.images.fixed_height.webp,
        img.images.fixed_height.mp4,
        img.images.fixed_height.url,
      ]
    : [
        img.images.fixed_width.webp,
        img.images.fixed_width.mp4,
        img.images.fixed_width.url,
      ];
}

export function getThumbnailUrl(img: GIFObject, layoutMode: LayoutMode) {
  return layoutMode === 'single_column'
    ? img.images.fixed_height_still.url
    : img.images.fixed_width_still.url;
}

export function getImageDimensions(img: GIFObject, layoutMode: LayoutMode) {
  const selected =
    layoutMode === 'single_column'
      ? img.images.fixed_height
      : img.images.fixed_width;

  return {
    width: parseInt(selected.width, 10) || 0,
    height: parseInt(selected.height, 10) || 0,
  };
}
