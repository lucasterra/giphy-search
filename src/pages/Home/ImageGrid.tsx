import React, { memo } from 'react';
import { getPlaceholderColor } from './placeholderColor';
import { getMainUrl, getThumbnailUrl, getImageDimensions } from './imageUrls';
import {
  ImageGrid as Grid,
  ProgressiveImage,
} from '../../components/ImageGrid';
import { LayoutMode } from './LayoutSwitcher';
import { GIFObject } from '../../api/GIPHY';

interface ImageGridProps {
  layoutMode: LayoutMode;
  images: GIFObject[];
}

export const ImageGrid = memo<ImageGridProps>(({ layoutMode, images }) => {
  let heights: number[] = [];
  let children: React.ReactNode[] = [];

  images.forEach((img, idx) => {
    const dimensions = getImageDimensions(img, layoutMode);

    heights.push(dimensions.height);
    children.push(
      <ProgressiveImage
        key={img.id}
        alt={img.title}
        thumbSrc={getThumbnailUrl(img, layoutMode)}
        src={getMainUrl(img, layoutMode)}
        backgroundColor={getPlaceholderColor(idx)}
        width={dimensions.width}
        height={dimensions.height}
      />
    );
  });

  return (
    <Grid
      numOfColumns={layoutMode === 'single_column' ? 1 : 3}
      childrenHeights={heights}
      children={children}
    />
  );
});
