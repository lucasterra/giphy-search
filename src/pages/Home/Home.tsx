import React, { useState } from 'react';
import { Layout } from '../../components/Layout';
import { Input } from '../../components/Input';
import { useImageSearch } from './useImageSearch';
import { LayoutMode, LayoutSwitcher } from './LayoutSwitcher';
import { InfiniteScrollButton } from './InfiniteScrollButton';
import { ImageGrid } from './ImageGrid';

const IMAGES_PER_PAGE = 24;

interface HomeProps {
  debounceTime?: number;
}

export const Home: React.FC<HomeProps> = ({ debounceTime = 250 }) => {
  const {
    text,
    setSearchTerm,
    loadMore,
    isLoading,
    imageData,
    hasMoreImages,
  } = useImageSearch(IMAGES_PER_PAGE, debounceTime);
  const [layoutMode, setLayoutMode] = useState<LayoutMode>('three_column');

  return (
    <Layout>
      <Input
        autoFocus
        data-testid="searchbox"
        placeholder="Type something to see some awesomeness"
        value={text}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          console.log('calling this!');
          setSearchTerm(e.target.value);
        }}
      />
      <LayoutSwitcher setLayoutMode={setLayoutMode} layoutMode={layoutMode} />

      <ImageGrid
        images={imageData && imageData.data ? imageData.data : []}
        layoutMode={layoutMode}
      />

      <InfiniteScrollButton
        data-testid="infinite-scroll-button"
        loadMore={loadMore}
        hidden={!hasMoreImages}
        isLoading={isLoading}
      />
    </Layout>
  );
};
