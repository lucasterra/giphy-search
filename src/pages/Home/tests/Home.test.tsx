import React from 'react';
import { render, fireEvent, waitForDomChange } from '@testing-library/react';
import { Home } from '../Home';

import trendingFirstPage from './fixtures/trending_0.json';
import trendingSecondPage from './fixtures/trending_1.json';
import kittenFirstPage from './fixtures/kitten_0.json';
import kittenSecondPage from './fixtures/kitten_1.json';

const mockFetch = jest.fn();
jest.spyOn(global as any, 'fetch').mockImplementation((url: string) => {
  mockFetch(url);

  switch (url) {
    case 'https://api.giphy.com/v1/gifs/trending?limit=24&offset=0&rating=G&lang=en&api_key=CdRKiCMbTnt9CkZTZ0lGukSczk6iT4Z6':
      return Promise.resolve({
        json: () => Promise.resolve(trendingFirstPage),
      });
    case 'https://api.giphy.com/v1/gifs/trending?limit=24&offset=24&rating=G&lang=en&api_key=CdRKiCMbTnt9CkZTZ0lGukSczk6iT4Z6':
      return Promise.resolve({
        json: () => Promise.resolve(trendingSecondPage),
      });
    case 'https://api.giphy.com/v1/gifs/search?q=kitten&limit=24&offset=0&rating=G&lang=en&api_key=CdRKiCMbTnt9CkZTZ0lGukSczk6iT4Z6':
      return Promise.resolve({
        json: () => Promise.resolve(kittenFirstPage),
      });
    case 'https://api.giphy.com/v1/gifs/search?q=kitten&limit=24&offset=24&rating=G&lang=en&api_key=CdRKiCMbTnt9CkZTZ0lGukSczk6iT4Z6':
      return Promise.resolve({
        json: () => Promise.resolve(kittenSecondPage),
      });
    default:
      return Promise.resolve({
        json: () => Promise.resolve({}),
      });
  }
});

describe('<Home />', () => {
  test('it renders correctly', async () => {
    // initial render
    const { getByTestId, queryByTestId } = render(<Home debounceTime={0} />);

    await waitForDomChange({
      container: getByTestId('infinite-scroll-button'),
    });

    expect(getByTestId('column-0')).toMatchSnapshot();
    expect(getByTestId('column-1')).toMatchSnapshot();
    expect(getByTestId('column-2')).toMatchSnapshot();

    // load more trending pictures
    fireEvent.click(getByTestId('infinite-scroll-button'));

    await waitForDomChange({
      container: getByTestId('infinite-scroll-button'),
    });

    expect(getByTestId('column-0')).toMatchSnapshot();
    expect(getByTestId('column-1')).toMatchSnapshot();
    expect(getByTestId('column-2')).toMatchSnapshot();

    // load kitten pictures :D
    fireEvent.change(getByTestId('searchbox'), {
      target: { value: 'kitten' },
    });

    await waitForDomChange({
      container: getByTestId('infinite-scroll-button'),
    });

    expect(getByTestId('column-0')).toMatchSnapshot();
    expect(getByTestId('column-1')).toMatchSnapshot();
    expect(getByTestId('column-2')).toMatchSnapshot();
  });
});
