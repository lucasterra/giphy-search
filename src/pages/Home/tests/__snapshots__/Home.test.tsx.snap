// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> it renders correctly 1`] = `
<div
  class="css-l2pxcp"
  data-testid="column-0"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 92.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(67, 41, 173);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 2`] = `
<div
  class="css-l2pxcp"
  data-testid="column-1"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100.5%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 83.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 83%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 55.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 3`] = `
<div
  class="css-l2pxcp"
  data-testid="column-2"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 104.5%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 51%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 61.5%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 48.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 76.5%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 89.5%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(51, 93, 173);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 4`] = `
<div
  class="css-l2pxcp"
  data-testid="column-0"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 92.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 162.5%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 68%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 70%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(159, 90, 87);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 5`] = `
<div
  class="css-l2pxcp"
  data-testid="column-1"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100.5%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 83.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 83%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 55.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 66.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 6`] = `
<div
  class="css-l2pxcp"
  data-testid="column-2"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 104.5%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 51%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 61.5%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 48.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 76.5%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 89.5%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 84%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 56.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 74%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(67, 41, 173);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 7`] = `
<div
  class="css-l2pxcp"
  data-testid="column-0"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 78%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 98%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 149%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 82%; background-color: rgb(173, 121, 5);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 74%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 94.5%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 74%; background-color: rgb(96, 163, 114);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 8`] = `
<div
  class="css-l2pxcp"
  data-testid="column-1"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 102.5%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 80%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 67%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 100%; background-color: rgb(196, 183, 210);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 75%; background-color: rgb(174, 41, 85);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 52.5%; background-color: rgb(96, 163, 114);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 82.5%; background-color: rgb(135, 202, 191);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 80%; background-color: rgb(159, 90, 87);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 125%; background-color: rgb(51, 93, 173);"
    />
  </div>
</div>
`;

exports[`<Home /> it renders correctly 9`] = `
<div
  class="css-l2pxcp"
  data-testid="column-2"
>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 80%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 107%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 94.5%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 133.5%; background-color: rgb(67, 41, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 88.5%; background-color: rgb(51, 93, 173);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 68%; background-color: rgb(242, 169, 112);"
    />
  </div>
  <div
    class="css-fy58gd"
  >
    <div
      class="css-8atqhb"
      data-testid="placeholder"
      style="padding-top: 125%; background-color: rgb(67, 41, 173);"
    />
  </div>
</div>
`;
